File: cpp/bucket_sort.cpp
// bucket sort algorithm

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

void bucketSort(vector<float>& arr) {
    int n = arr.size();
    vector<vector<float>> buckets(n);
    for(int i=0; i<n; i++) {
        int idx = n * arr[i];
        buckets[idx].push_back(arr[i]);
    }
    for(int i=0; i<n; i++) sort(buckets[i].begin(), buckets[i].end());
    int idx = 0;
    for(int i=0; i<n; i++) {
        for(float x : buckets[i]) arr[idx++] = x;
    }
}

int main() {
    vector<float> arr = {0.897, 0.565, 0.656, 0.123, 0.665, 0.343};
    bucketSort(arr);
    for(float x : arr) cout << x << " ";
    return 0;
}

File: python/bucket_sort.py
# bucket sort algorithm 

def bucket_sort(arr):
    n = len(arr)
    buckets = [[] for _ in range(n)]
    for num in arr:
        idx = int(num * n)
        buckets[idx].append(num)
    for i in range(n):
        buckets[i].sort()
    result = []
    for bucket in buckets:
        result.extend(bucket)
    return result

arr = [0.897, 0.565, 0.656, 0.123, 0.665, 0.343]
print(bucket_sort(arr))
